sql-plugin/src/main/scala/com/nvidia/spark/rapids/FoldLocalAggregate.scala

```diff
@@ -0,0 +1,137 @@
+        }
+    }
+  }
+}
+
+/**
+ * Capture the pattern of LocalAggregate, which consists of two-stage AggregateExecs being
+ * connected with each other directly (without the ShuffleExchange in between):
+ * +- FinalAggregateExec
+ *    +- PartialAggregateExec
+ *
+ * The LocalAggregate can be emerged regardless HashAggregateExec, SortAggregateExec or
+ * ObjectHashAggregateExec.
+ */
+object LocalAggregatePattern extends Logging {
+  def unapply(plan: SparkPlan): Option[(BaseAggregateExec, BaseAggregateExec)] = {
+    plan match {
+      case hashAgg: HashAggregateExec
+        if hashAgg.child.isInstanceOf[HashAggregateExec] &&
+          isLocalExchange(hashAgg, hashAgg.child.asInstanceOf[HashAggregateExec]) =>
+        Some(hashAgg -> hashAgg.child.asInstanceOf[HashAggregateExec])
+      case sort: SortAggregateExec
+        if sort.child.isInstanceOf[SortAggregateExec] &&
+          isLocalExchange(sort, sort.child.asInstanceOf[SortAggregateExec]) =>
+        Some(sort -> sort.child.asInstanceOf[SortAggregateExec])
+      case obj: ObjectHashAggregateExec
+        if obj.child.isInstanceOf[ObjectHashAggregateExec] &&
+          isLocalExchange(obj, obj.child.asInstanceOf[ObjectHashAggregateExec])  =>
+        Some(obj -> obj.child.asInstanceOf[ObjectHashAggregateExec])
+      case _ =>
+        None
+    }
+  }
+
+  // Check if the two connected AggregateExecs are two stages of the same logical Aggregate. If
+  // so, group expressions and aggregate expressions should be identical(except AggregateMode)
+  private def isLocalExchange(merge: BaseAggregateExec, partial: BaseAggregateExec): Boolean = {
+    val groupExpressions = merge.groupingExpressions
+    val aggExpressions = merge.aggregateExpressions
+    val childGroupExpressions = partial.groupingExpressions
+    val childAggExpressions = partial.aggregateExpressions
+    val requiredDistribution = merge.requiredChildDistribution
+
+    if ( // Fast check
+      groupExpressions.length != childGroupExpressions.length ||
+        aggExpressions.length != childAggExpressions.length) {
+      false
+    } else if ( // Check if partial OutputPartition statisfy distribution of FinalAgg
+      !partial.outputPartitioning.satisfies(requiredDistribution.head)) {
+      false
+    } else if ( // Check AggregateExpressions
+      !aggExpressions.zip(childAggExpressions).forall {
+        // <Partial -> Final> pair check
+        case (s2, s1) if s2.mode != Final || s1.mode != Partial =>
+          false
+        // TODO: Currently, distinct aggregate is NOT supported
+        case (s2, s1) if s2.isDistinct || s1.isDistinct =>
+          false
+        // AggregateFunctions should be identical
+        case (s2, s1) =>
+          s2.aggregateFunction.equals(s1.aggregateFunction)
+      }) {
+      false
+    } else {
+      // Check GroupExpressions
+      groupExpressions.zip(childGroupExpressions).forall {
+        case (s2, s1) => s2.semanticEquals(s1.toAttribute)
```

Comment:
Why only convert s1 to an attribute ?

Reply:
Cuz groupExpressions of FinalAggregate are GroupAttributes: https://github.com/apache/spark/blob/branch-3.5/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggUtils.scala#L162